File: BaekJoon Online Judge/C++/Step18-Backtracking/15649.cpp
//
//  15649.cpp
//  Step18-Backtracking
//
//  Created by 김예빈 on 2019. 9. 21..
//  Copyright © 2019년 김예빈. All rights reserved.
//

#include <iostream>
#include <vector>
using namespace std;

int N, M;
vector<int> graph;
bool visit[9];

void dfs(int, vector<int>, bool[]);

// permutation
int main(int argc, const char * argv[]) {
    cin.tie(NULL);
    ios::sync_with_stdio(false);
    
    cin >> N >> M;
    
    dfs(0, graph, visit);
    
    return 0;
}

void dfs(int start, vector<int> graph, bool visit[]) {
    if(start == M) {
        for(int i = 0; i < graph.size(); i++)
            cout << graph[i] << " ";
        
        cout << "\n";
        return;
    }
    
    for (int i = 1; i <= N; i++) {
        if (visit[i] == true) continue;
        
        visit[i] = true;
        graph.push_back(i);
        
        dfs(start + 1, graph, visit);
        
        visit[i] = false;
        graph.pop_back();
    }
}

File: BaekJoon Online Judge/Swift/Step21-Backtracking/15649.swift
//
//  15649.swift
//  Step13-Backtracking
//
//  Created by Yebin Kim on 2019/12/3.
//  Copyright © 2019 김예빈. All rights reserved.
//

import Foundation

let input = readLine()!.split(separator: " ").map{ Int($0)! }
let N = input[0]
let M = input[1]

var visit = Array<Bool>(repeating: false, count: 9)
var graph = Array<Int>()

func dfs(_ start: Int) {
    if start == M {
        for i in 0..<graph.count {
            print(graph[i], terminator: " ")
        }
        
        print()
        return
    }
    
    for i in 1...N {
        if visit[i] == true {
            continue
        }
        
        visit[i] = true
        graph.append(i)
        
        dfs(start + 1)
        
        visit[i] = false
        graph.removeLast()
    }
}

dfs(0)
